heareval/predictions/task_predictions.py

```diff
@@ -0,0 +1,127 @@
+        super().__init__()
+
+        self.projection = torch.nn.Linear(nfeatures, nlabels)
+        if prediction_type == "multilabel":
+            self.activation = torch.nn.Sigmoid()
+        elif prediction_type == "multiclass":
+            self.activation = torch.nn.Softmax()
+        else:
+            raise ValueError(f"Unknown prediction_type {prediction_type}")
+
+    def forward(self, x: torch.Tensor):
+        x = self.projection(x)
+        x = self.activation(x)
+        return x
+
+
+class SplitMemmapDataset(Dataset):
+    """
+    Embeddings are memmap'ed. (We don't use the labels yet, but will add them later.)
+
+    WARNING: Don't shuffle this or access will be SLOW.
+    """
+
+    def __init__(self, embedding_path: Path, split_name: str):
+        self.embedding_path = embedding_path
+        self.split_name = split_name
+
+        self.dim = tuple(
+            json.load(
+                open(embedding_path.joinpath(f"{split_name}.embedding-dimensions.json"))
+            )
+        )
+        self.embedding_memmap = np.memmap(
+            filename=embedding_path.joinpath(f"{split_name}.embeddings.npy"),
+            dtype=np.float32,
+            mode="r",
+            shape=self.dim,
+        )
+        # self.labels = pickle.load(open(embedding_path.joinpath(f"{split_name}.target-labels.pkl"), "rb"))
+        # assert len(self.labels) == dim[0]
+
+    def __len__(self):
+        return self.dim[0]
+        # return len(self.labels)
+
+    def __getitem__(self, idx):
+        return self.embedding_memmap[idx]
+        # return self.embedding_memmap[idx], self.labels[idx]
+
+
+def task_predictions(
+    embedding_path: Path, scene_embedding_size: int, timestamp_embedding_size: int
+):
+    metadata = json.load(embedding_path.joinpath("task_metadata.json").open())
+
+    label_vocab = pd.read_csv(embedding_path.joinpath("labelvocabulary.csv"))
+
+    nlabels = len(label_vocab)
+    assert nlabels == label_vocab["idx"].max() + 1
+
+    if metadata["embedding_type"] == "scene":
+        predictor = RandomProjectionPrediction(
+            scene_embedding_size, nlabels, metadata["prediction_type"]
+        )
+    elif metadata["embedding_type"] == "event":
+        predictor = RandomProjectionPrediction(
+            timestamp_embedding_size, nlabels, metadata["prediction_type"]
+        )
+    else:
+        raise ValueError(f"embedding_type {metadata['embedding_type']} unknown.")
+
+    # for split in metadata["splits"]:
+    # Only do test for now, for the unlearned random predictor
+    for split in [{"name": "test"}]:
+        print(f"Getting embeddings for split: {split['name']}")
+
+        dataloader = DataLoader(
+            SplitMemmapDataset(embedding_path, split["name"]),
+            batch_size=64,
+            # We don't shuffle because it's slow.
+            # Also we want predicted labels in the same order as
+            # target labels.
```

Comment:
Is the issue that reading adjacent items using memmap is significantly faster? Also -- I think if we using the dataset to return target labels then we don't need to worry about the ordering here.

Reply:
@jorshi we're going to train in this order too.

Random access to a memmap is slow because you're seeking around disk